Create an example for sampling multiple chains using vmap and pmap (#273)
blackjax-devs/blackjax: examples/MultipleChains.md

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.14.1
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

# Sampling Multiple Chains

> In this example, we'll briefly demonstrate how you can run multiple MCMC chains using `jax` built-in constructs: `vmap` and `pmap`.

Reviewer: ```suggestion
In this example, we will briefly demonstrate how you can run multiple MCMC chains using `jax` built-in constructs: `vmap` and `pmap`.
```

> We will use the NUTS example from the previous notebook, and compare the performance of the two approaches.

Reviewer: ```suggestion
We will use the NUTS example from the introduction notebook, and compare the performance of the two approaches.
```

## Vectorization vs parallelization

`jax` provides two distinct transformations:
- [vmap](https://jax.readthedocs.io/en/latest/jax.html?highlight=vmap#jax.vmap), used to automatically vectorize `jax` code
- and [pmap](https://jax.readthedocs.io/en/latest/_autosummary/jax.pmap.html#jax.pmap),

> which enables paralleiziation across multiple devices, such as multiple GPUs (or, in our case, CPU cores).

Reviewer: ```suggestion
which enables parallelization across multiple devices, such as multiple GPUs (or, in our case, CPU cores).
```

Unfortunately, this means that using `pmap` is not possible out of the box -- we'll first need
to instruct `jax` to split the CPU into multiple devices.
Please see [this issue](https://github.com/google/jax/issues/1408) for more discussion on this topic.

Currently, this can only be done via `XLA_FLAGS` environmental variable.

**Note that this variable has to be set before any `jax` code is executed**

```{code-cell} ipython3
import os
import multiprocessing

os.environ["XLA_FLAGS"] = "--xla_force_host_platform_device_count={}".format(
    multiprocessing.cpu_count()
)
```

We can now import `jax` and confirm that it have successfuly recognized our CPU as multiple devices.

```{code-cell} ipython3
import jax
import jax.numpy as jnp

len(jax.devices())
```

```{code-cell} ipython3
jax.devices()[:2]
```

### Choosing the number of devices

`pmap` has one more limitation - it is not able to parallelize the execution when the number of items is larger than the number of devices.

```{code-cell} ipython3
def fn(x):
    return x + 1

try:
    data = jnp.arange(1024)
    parallel_fn = jax.pmap(fn)

    parallel_fn(data)

except Exception as e:
    print(e)
```

> This means that you'll only be able to run as many MCMC chains as you have CPU cores.

Reviewer: ```suggestion
This means that you will only be able to run as many MCMC chains as you have CPU cores.
```

> See this [question](https://github.com/google/jax/discussions/4198) for a more discussion on the topic,

Reviewer: ```suggestion
See this [question](https://github.com/google/jax/discussions/4198) for a more detailed discussion on the topic,
```

```{code-cell} ipython3
initial_positions = {"loc": np.ones(num_chains), "scale": 2.0 * np.ones(num_chains)}
initial_states = jax.vmap(nuts.init, in_axes=(0))(initial_positions)
```

And finally run the sampler

```{code-cell} ipython3
%%time

states = inference_loop_multiple_chains(
    rng_key, nuts.step, initial_states, 2_000, num_chains
)
_ = states.position["loc"].block_until_ready()
```

You can now access the samples from individual chains by simply indexing the returned arrays:

```{code-cell} ipython3
states.position["loc"].shape
```

E.g. to get the `loc` samples for the second chain:

```{code-cell} ipython3
samples = states.position["loc"][:, 1]
samples
```

```{code-cell} ipython3
samples.shape
```

```{code-cell} ipython3
fig, ax = plt.subplots(figsize=(15, 6))

ax.plot(states.position["loc"][:, 0], color="blue", alpha=0.25)
ax.plot(states.position["loc"][:, 1], color="red", alpha=0.25)
ax.set_xlabel("Samples")
ax.set_ylabel("loc")
ax.legend(["Chain 1", "Chain 2"])
```

### Using `pmap`

In case of `pmap`, we can simply choose to apply the transformation directly to the original `inference_loop` function.

```{code-cell} ipython3

> inference_loop_multiple_chains = jax.pmap(inference_loop, in_axes=(0, None, 0, None), static_broadcasted_argnums=(1, 3))

Reviewer: What happens if you apply `jax.vmap` to `inference_loop`?

Author: Performance-wise it's exactly the same on my machine. 
IMO the code is then a bit cleaner (no need to write a new function).

Should I change the example and apply `vmap` to `inference_loop`? I didn't change the original example since I thought there might be some other reasons to do manually implement `inference_loop_multiple_chains`
